File: packages/core/src/consultation/dialogue-manager.ts
/**
 * Dialogue File Management for Story 86-3
 *
 * Persistence layer for tandem agent consultation exchanges.
 * Creates, appends, and archives dialogue files following the format
 * defined in ADR-0012 (lines 156-195).
 *
 * All functions are pure or use simple file I/O — no external dependencies.
 */

import fs from 'node:fs';
import path from 'node:path';

// =============================================================================
// Types
// =============================================================================

export type Outcome = 'applied' | 'deferred' | 'rejected';

export interface DialogueHeader {
  storyId: string;
  workflow: string;
  leader: string;
  leaderCharacter?: string;
  partner: string;
  partnerCharacter?: string;
  startedAt: string;
}

export interface DialogueExchange {
  number: number;
  timestamp: string; // HH:MM
  leader: string;
  partner: string;
  question: string;
  recommendation: string;
  confidence: string;
  outcome?: Outcome;
  outcomeNote?: string;
}

export interface DialogueResult {
  success: boolean;
  data?: Record<string, unknown>;
  error?: string;
}

interface ArchiveParams {
  dialoguePath: string;
  archiveDir: string;
  jiraKey?: string;
  storyId?: string;
}

// =============================================================================
// Constants
// =============================================================================

const SUMMARY_MARKER = '## Summary';
const EXCHANGE_RE = /^## Exchange (\d+)/;
const OUTCOME_RE = /^\*\*Outcome:\*\*\s+(.+)/;
const DIRECTION_RE = /^\*\*\[(\d{2}:\d{2})\]\s+(\S+)\s+→\s+(\S+)\*\*/;
const PARTNER_RESP_RE = /^\*\*\[(\d{2}:\d{2})\]\s+(\S+):\*\*/;
const CONFIDENCE_RE = /^\*\*Confidence:\*\*\s+(\S+)/;

// =============================================================================
// Pure Functions
// =============================================================================

/**
 * Create initial dialogue file content with header and empty summary.
 */
export function createDialogueContent(header: DialogueHeader): string {
  const leaderLabel = header.leaderCharacter
    ? `${header.leader} (${header.leaderCharacter})`
    : header.leader;
  const partnerLabel = header.partnerCharacter
    ? `${header.partner} (${header.partnerCharacter})`
    : header.partner;

  return `# Tandem Dialogue: ${header.storyId}

**Workflow:** ${header.workflow}
**Leader:** ${leaderLabel} | **Partner:** ${partnerLabel}
**Started:** ${header.startedAt}

---

${SUMMARY_MARKER}
- **Total exchanges:** 0
- **Key decisions:** None
- **Time in tandem:** 0m
`;
}

/**
 * Format a single exchange as markdown block.
 */
export function formatExchange(exchange: DialogueExchange): string {
  const outcomeText = exchange.outcome
    ? `**Outcome:** ${exchange.outcome}${exchange.outcomeNote ? ` - ${exchange.outcomeNote}` : ''}`
    : '**Outcome:** _pending_';

  return `## Exchange ${exchange.number}
**[${exchange.timestamp}] ${exchange.leader} → ${exchange.partner}**

> ${exchange.question}

**[${exchange.timestamp}] ${exchange.partner}:**

${exchange.recommendation}

**Confidence:** ${exchange.confidence}

${outcomeText}

---
`;
}

/**
 * Parse exchanges from dialogue file content.
 */
export function parseDialogueExchanges(content: string): DialogueExchange[] {
  const exchanges: DialogueExchange[] = [];
  const lines = content.split('\n');

  let current: Partial<DialogueExchange> | null = null;
  let inQuestion = false;
  let inRecommendation = false;
  let questionLines: string[] = [];
  let recLines: string[] = [];

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    // New exchange starts
    const exchangeMatch = line.match(EXCHANGE_RE);
    if (exchangeMatch) {
      if (current && current.number !== undefined) {
        current.question = questionLines.join('\n');
        current.recommendation = recLines.join('\n');
        exchanges.push(current as DialogueExchange);
      }
      current = { number: parseInt(exchangeMatch[1], 10) };
      questionLines = [];
      recLines = [];
      inQuestion = false;
      inRecommendation = false;
      continue;
    }

    if (!current) continue;

    // Leader → Partner direction line
    const dirMatch = line.match(DIRECTION_RE);
    if (dirMatch) {
      current.timestamp = dirMatch[1];
      current.leader = dirMatch[2];
      current.partner = dirMatch[3];
      inQuestion = true;
      inRecommendation = false;
      continue;
    }

    // Partner response line
    const partnerMatch = line.match(PARTNER_RESP_RE);
    if (partnerMatch) {
      inQuestion = false;
      inRecommendation = true;
      continue;
    }

    // Confidence line
    const confMatch = line.match(CONFIDENCE_RE);
    if (confMatch) {
      current.confidence = confMatch[1];
      inRecommendation = false;
      continue;
    }

    // Outcome line
    const outcomeMatch = line.match(OUTCOME_RE);
    if (outcomeMatch) {
      const outcomeRaw = outcomeMatch[1].trim();
      if (outcomeRaw === '_pending_') {
        // No outcome yet
      } else {
        const dashIdx = outcomeRaw.indexOf(' - ');
        if (dashIdx >= 0) {
          current.outcome = outcomeRaw.substring(0, dashIdx).trim() as Outcome;
          current.outcomeNote = outcomeRaw.substring(dashIdx + 3).trim();
        } else {
          current.outcome = outcomeRaw as Outcome;
        }
      }
      inRecommendation = false;
      continue;
    }

    // Collect question text (blockquote lines)
    if (inQuestion) {
      const stripped = line.startsWith('> ') ? line.substring(2) : line;
      if (stripped.trim()) {
        questionLines.push(stripped);
      }
      continue;
    }

    // Collect recommendation text
    if (inRecommendation) {
      if (line.trim()) {
        recLines.push(line);
      }
      continue;
    }
  }

  // Push last exchange
  if (current && current.number !== undefined) {
    current.question = questionLines.join('\n');
    current.recommendation = recLines.join('\n');
    exchanges.push(current as DialogueExchange);
  }

  return exchanges;
}

/**
 * Generate summary markdown section from exchanges.
 */
export function generateSummary(exchanges: DialogueExchange[], _startedAt: string): string {
  const total = exchanges.length;

  // Key decisions from applied outcomes
  const applied = exchanges.filter(e => e.outcome === 'applied' && e.outcomeNote);
  const decisionsText = applied.length > 0
    ? applied.map(e => `  - ${e.outcomeNote}`).join('\n')
    : 'None';

  // Time in tandem: span between first and last exchange timestamps
  let duration = '0m';
  if (exchanges.length > 0) {
    const first = parseTime(exchanges[0].timestamp);
    const last = parseTime(exchanges[exchanges.length - 1].timestamp);
    if (first !== null && last !== null) {
      const mins = last - first;
      duration = mins > 0 ? `${mins}m` : '0m';
    }
  }

  return `${SUMMARY_MARKER}
- **Total exchanges:** ${total}
- **Key decisions:**
${decisionsText}
- **Time in tandem:** ${duration}
`;
}

// =============================================================================
// File Operations
// =============================================================================

/**
 * Append an exchange to a dialogue file. Creates the file if missing.
 */
export async function appendExchangeToFile(
  dialoguePath: string,
  exchange: DialogueExchange,
  header?: DialogueHeader,
): Promise<DialogueResult> {
  try {
    if (!fs.existsSync(dialoguePath)) {
      if (!header) {
        return { success: false, error: 'Header required for new dialogue file' };
      }
      const initial = createDialogueContent(header);
      fs.mkdirSync(path.dirname(dialoguePath), { recursive: true });
      fs.writeFileSync(dialoguePath, initial, 'utf-8');
    }

    const content = fs.readFileSync(dialoguePath, 'utf-8');
    const formatted = formatExchange(exchange);

    // Insert exchange before summary section
    const summaryIdx = content.indexOf(SUMMARY_MARKER);
    if (summaryIdx < 0) {
      // No summary marker — append at end
      fs.writeFileSync(dialoguePath, content + '\n' + formatted, 'utf-8');
    } else {
      const before = content.substring(0, summaryIdx);
      const after = content.substring(summaryIdx);
      fs.writeFileSync(dialoguePath, before + formatted + '\n' + after, 'utf-8');
    }

    return { success: true, data: { exchangeNumber: exchange.number } };
  } catch (err) {
    return { success: false, error: `Failed to append exchange: ${err instanceof Error ? err.message : String(err)}` };
  }
}

/**
 * Update the outcome of a specific exchange in the dialogue file.
 */
export async function updateOutcomeInFile(
  dialoguePath: string,
  exchangeNum: number,
  outcome: Outcome,
  note?: string,
): Promise<DialogueResult> {
  try {
    if (!fs.existsSync(dialoguePath)) {
      return { success: false, error: `Dialogue file not found: ${dialoguePath}` };
    }

    const content = fs.readFileSync(dialoguePath, 'utf-8');
    const lines = content.split('\n');

    // Find the exchange section
    let inTargetExchange = false;
    let found = false;

    for (let i = 0; i < lines.length; i++) {
      const exchangeMatch = lines[i].match(EXCHANGE_RE);
      if (exchangeMatch) {
        inTargetExchange = parseInt(exchangeMatch[1], 10) === exchangeNum;
      }

      if (inTargetExchange && OUTCOME_RE.test(lines[i])) {
        const outcomeText = note
          ? `**Outcome:** ${outcome} - ${note}`
          : `**Outcome:** ${outcome}`;
        lines[i] = outcomeText;
        found = true;
        break;
      }
    }

    if (!found) {
      return { success: false, error: `Exchange ${exchangeNum} not found in dialogue file` };
    }

    fs.writeFileSync(dialoguePath, lines.join('\n'), 'utf-8');
    return { success: true, data: { exchangeNum, outcome } };
  } catch (err) {
    return { success: false, error: `Failed to update outcome: ${err instanceof Error ? err.message : String(err)}` };
  }
}

/**
 * Regenerate the summary section in an existing dialogue file.
 */
export async function refreshSummary(dialoguePath: string): Promise<DialogueResult> {
  try {
    if (!fs.existsSync(dialoguePath)) {
      return { success: false, error: `Dialogue file not found: ${dialoguePath}` };
    }

    const content = fs.readFileSync(dialoguePath, 'utf-8');
    const exchanges = parseDialogueExchanges(content);

    // Extract startedAt from header
    const startedMatch = content.match(/\*\*Started:\*\*\s+(.+)/);
    const startedAt = startedMatch ? startedMatch[1].trim() : new Date().toISOString();

    const newSummary = generateSummary(exchanges, startedAt);

    // Replace existing summary section
    const summaryIdx = content.indexOf(SUMMARY_MARKER);
    if (summaryIdx < 0) {
      // Append summary at end
      fs.writeFileSync(dialoguePath, content + '\n' + newSummary, 'utf-8');
    } else {
      const before = content.substring(0, summaryIdx);
      fs.writeFileSync(dialoguePath, before + newSummary, 'utf-8');
    }

    return { success: true, data: { totalExchanges: exchanges.length } };
  } catch (err) {
    return { success: false, error: `Failed to refresh summary: ${err instanceof Error ? err.message : String(err)}` };
  }
}

/**
 * Copy dialogue file to archive directory.
 */
export async function archiveDialogue(params: ArchiveParams): Promise<DialogueResult> {
  const { dialoguePath, archiveDir, jiraKey, storyId } = params;

  try {
    if (!fs.existsSync(dialoguePath)) {
      return { success: false, error: `Dialogue file not found: ${dialoguePath}` };
    }

    fs.mkdirSync(archiveDir, { recursive: true });

    const prefix = jiraKey || storyId || 'unknown';
    const archiveName = `${prefix}-dialogue.md`;
    const archivePath = path.join(archiveDir, archiveName);

    fs.copyFileSync(dialoguePath, archivePath);

    return { success: true, data: { archivePath } };
  } catch (err) {
    return { success: false, error: `Failed to archive dialogue: ${err instanceof Error ? err.message : String(err)}` };
  }
}

// =============================================================================
// Internal Helpers
// =============================================================================

/**
 * Parse HH:MM timestamp to minutes since midnight.
 */
function parseTime(timestamp: string): number | null {
  const parts = timestamp.split(':');
  if (parts.length !== 2) return null;
  const hours = parseInt(parts[0], 10);
  const minutes = parseInt(parts[1], 10);
  if (Number.isNaN(hours) || Number.isNaN(minutes)) return null;
  return hours * 60 + minutes;
}

File: runtime/src/pf/consultation/dialogue_manager.py
"""Dialogue file management for tandem agent consultation.

Persistence layer for consultation exchanges between tandem agents.
Port of packages/core/src/consultation/dialogue-manager.ts to Python.

All pure functions use the same markdown format defined in ADR-0012.
"""

from __future__ import annotations

import re
import shutil
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

Outcome = Literal["applied", "deferred", "rejected"]

SUMMARY_MARKER = "## Summary"
EXCHANGE_RE = re.compile(r"^## Exchange (\d+)")
OUTCOME_RE = re.compile(r"^\*\*Outcome:\*\*\s+(.+)")
DIRECTION_RE = re.compile(r"^\*\*\[(\d{2}:\d{2})\]\s+(\S+)\s+→\s+(\S+)\*\*")
PARTNER_RESP_RE = re.compile(r"^\*\*\[(\d{2}:\d{2})\]\s+(\S+):\*\*")
CONFIDENCE_RE = re.compile(r"^\*\*Confidence:\*\*\s+(\S+)")


@dataclass
class DialogueHeader:
    story_id: str
    workflow: str
    leader: str
    partner: str
    leader_character: str | None = None
    partner_character: str | None = None
    started_at: str = ""


@dataclass
class DialogueExchange:
    number: int
    timestamp: str  # HH:MM
    leader: str
    partner: str
    question: str
    recommendation: str
    confidence: str
    outcome: Outcome | None = None
    outcome_note: str | None = None


@dataclass
class DialogueResult:
    success: bool
    data: dict | None = None
    error: str | None = None


# =============================================================================
# Pure Functions
# =============================================================================


def create_dialogue_content(header: DialogueHeader) -> str:
    """Create initial dialogue file content with header and empty summary."""
    leader_label = (
        f"{header.leader} ({header.leader_character})" if header.leader_character else header.leader
    )
    partner_label = (
        f"{header.partner} ({header.partner_character})"
        if header.partner_character
        else header.partner
    )

    return (
        f"# Tandem Dialogue: {header.story_id}\n"
        f"\n"
        f"**Workflow:** {header.workflow}\n"
        f"**Leader:** {leader_label} | **Partner:** {partner_label}\n"
        f"**Started:** {header.started_at}\n"
        f"\n"
        f"---\n"
        f"\n"
        f"{SUMMARY_MARKER}\n"
        f"- **Total exchanges:** 0\n"
        f"- **Key decisions:** None\n"
        f"- **Time in tandem:** 0m\n"
    )


def format_exchange(exchange: DialogueExchange) -> str:
    """Format a single exchange as markdown block."""
    if exchange.outcome:
        outcome_text = f"**Outcome:** {exchange.outcome}"
        if exchange.outcome_note:
            outcome_text += f" - {exchange.outcome_note}"
    else:
        outcome_text = "**Outcome:** _pending_"

    return (
        f"## Exchange {exchange.number}\n"
        f"**[{exchange.timestamp}] {exchange.leader} \u2192 {exchange.partner}**\n"
        f"\n"
        f"> {exchange.question}\n"
        f"\n"
        f"**[{exchange.timestamp}] {exchange.partner}:**\n"
        f"\n"
        f"{exchange.recommendation}\n"
        f"\n"
        f"**Confidence:** {exchange.confidence}\n"
        f"\n"
        f"{outcome_text}\n"
        f"\n"
        f"---\n"
    )


def parse_dialogue_exchanges(content: str) -> list[DialogueExchange]:
    """Parse exchanges from dialogue file content."""
    exchanges: list[DialogueExchange] = []
    lines = content.split("\n")

    current: dict | None = None
    in_question = False
    in_recommendation = False
    question_lines: list[str] = []
    rec_lines: list[str] = []

    for line in lines:
        # New exchange starts
        m = EXCHANGE_RE.match(line)
        if m:
            if current is not None and "number" in current:
                current["question"] = "\n".join(question_lines)
                current["recommendation"] = "\n".join(rec_lines)
                exchanges.append(DialogueExchange(**current))
            current = {"number": int(m.group(1))}
            question_lines = []
            rec_lines = []
            in_question = False
            in_recommendation = False
            continue

        if current is None:
            continue

        # Leader → Partner direction line
        dm = DIRECTION_RE.match(line)
        if dm:
            current["timestamp"] = dm.group(1)
            current["leader"] = dm.group(2)
            current["partner"] = dm.group(3)
            in_question = True
            in_recommendation = False
            continue

        # Partner response line
        pm = PARTNER_RESP_RE.match(line)
        if pm:
            in_question = False
            in_recommendation = True
            continue

        # Confidence line
        cm = CONFIDENCE_RE.match(line)
        if cm:
            current["confidence"] = cm.group(1)
            in_recommendation = False
            continue

        # Outcome line
        om = OUTCOME_RE.match(line)
        if om:
            outcome_raw = om.group(1).strip()
            if outcome_raw != "_pending_":
                dash_idx = outcome_raw.find(" - ")
                if dash_idx >= 0:
                    current["outcome"] = outcome_raw[:dash_idx].strip()
                    current["outcome_note"] = outcome_raw[dash_idx + 3 :].strip()
                else:
                    current["outcome"] = outcome_raw
            in_recommendation = False
            continue

        # Collect question text (blockquote lines)
        if in_question:
            stripped = line[2:] if line.startswith("> ") else line
            if stripped.strip():
                question_lines.append(stripped)
            continue

        # Collect recommendation text
        if in_recommendation:
            if line.strip():
                rec_lines.append(line)
            continue

    # Push last exchange
    if current is not None and "number" in current:
        current["question"] = "\n".join(question_lines)
        current["recommendation"] = "\n".join(rec_lines)
        exchanges.append(DialogueExchange(**current))

    return exchanges


def generate_summary(exchanges: list[DialogueExchange], started_at: str) -> str:
    """Generate summary markdown section from exchanges."""
    total = len(exchanges)

    # Key decisions from applied outcomes
    applied = [e for e in exchanges if e.outcome == "applied" and e.outcome_note]
    if applied:
        decisions_text = "\n".join(f"  - {e.outcome_note}" for e in applied)
    else:
        decisions_text = "None"

    # Time in tandem: span between first and last exchange timestamps
    duration = "0m"
    if exchanges:
        first = _parse_time(exchanges[0].timestamp)
        last = _parse_time(exchanges[-1].timestamp)
        if first is not None and last is not None:
            mins = last - first
            duration = f"{mins}m" if mins > 0 else "0m"

    return (
        f"{SUMMARY_MARKER}\n"
        f"- **Total exchanges:** {total}\n"
        f"- **Key decisions:**\n"
        f"{decisions_text}\n"
        f"- **Time in tandem:** {duration}\n"
    )


# =============================================================================
# File Operations
# =============================================================================


def append_exchange_to_file(
    dialogue_path: Path,
    exchange: DialogueExchange,
    header: DialogueHeader | None = None,
) -> DialogueResult:
    """Append an exchange to a dialogue file. Creates the file if missing."""
    try:
        if not dialogue_path.exists():
            if not header:
                return DialogueResult(success=False, error="Header required for new dialogue file")
            initial = create_dialogue_content(header)
            dialogue_path.parent.mkdir(parents=True, exist_ok=True)
            dialogue_path.write_text(initial, encoding="utf-8")

        content = dialogue_path.read_text(encoding="utf-8")
        formatted = format_exchange(exchange)

        # Insert exchange before summary section
        summary_idx = content.find(SUMMARY_MARKER)
        if summary_idx < 0:
            dialogue_path.write_text(content + "\n" + formatted, encoding="utf-8")
        else:
            before = content[:summary_idx]
            after = content[summary_idx:]
            dialogue_path.write_text(before + formatted + "\n" + after, encoding="utf-8")

        return DialogueResult(success=True, data={"exchangeNumber": exchange.number})
    except Exception as err:
        return DialogueResult(success=False, error=f"Failed to append exchange: {err}")


def update_outcome_in_file(
    dialogue_path: Path,
    exchange_num: int,
    outcome: Outcome,
    note: str | None = None,
) -> DialogueResult:
    """Update the outcome of a specific exchange in the dialogue file."""
    try:
        if not dialogue_path.exists():
            return DialogueResult(
                success=False,
                error=f"Dialogue file not found: {dialogue_path}",
            )

        content = dialogue_path.read_text(encoding="utf-8")
        lines = content.split("\n")

        in_target = False
        found = False

        for i, line in enumerate(lines):
            em = EXCHANGE_RE.match(line)
            if em:
                in_target = int(em.group(1)) == exchange_num

            if in_target and OUTCOME_RE.match(line):
                outcome_text = f"**Outcome:** {outcome}"
                if note:
                    outcome_text += f" - {note}"
                lines[i] = outcome_text
                found = True
                break

        if not found:
            return DialogueResult(
                success=False,
                error=f"Exchange {exchange_num} not found in dialogue file",
            )

        dialogue_path.write_text("\n".join(lines), encoding="utf-8")
        return DialogueResult(success=True, data={"exchangeNum": exchange_num, "outcome": outcome})
    except Exception as err:
        return DialogueResult(success=False, error=f"Failed to update outcome: {err}")


def refresh_summary(dialogue_path: Path) -> DialogueResult:
    """Regenerate the summary section in an existing dialogue file."""
    try:
        if not dialogue_path.exists():
            return DialogueResult(
                success=False,
                error=f"Dialogue file not found: {dialogue_path}",
            )

        content = dialogue_path.read_text(encoding="utf-8")
        exchanges = parse_dialogue_exchanges(content)

        # Extract startedAt from header
        started_match = re.search(r"\*\*Started:\*\*\s+(.+)", content)
        started_at = started_match.group(1).strip() if started_match else ""

        new_summary = generate_summary(exchanges, started_at)

        # Replace existing summary section
        summary_idx = content.find(SUMMARY_MARKER)
        if summary_idx < 0:
            dialogue_path.write_text(content + "\n" + new_summary, encoding="utf-8")
        else:
            before = content[:summary_idx]
            dialogue_path.write_text(before + new_summary, encoding="utf-8")

        return DialogueResult(success=True, data={"totalExchanges": len(exchanges)})
    except Exception as err:
        return DialogueResult(success=False, error=f"Failed to refresh summary: {err}")


def archive_dialogue(
    dialogue_path: Path,
    archive_dir: Path,
    jira_key: str | None = None,
    story_id: str | None = None,
) -> DialogueResult:
    """Copy dialogue file to archive directory."""
    try:
        if not dialogue_path.exists():
            return DialogueResult(
                success=False,
                error=f"Dialogue file not found: {dialogue_path}",
            )

        archive_dir.mkdir(parents=True, exist_ok=True)

        prefix = jira_key or story_id or "unknown"
        archive_name = f"{prefix}-dialogue.md"
        archive_path = archive_dir / archive_name

        shutil.copy2(dialogue_path, archive_path)

        return DialogueResult(success=True, data={"archivePath": str(archive_path)})
    except Exception as err:
        return DialogueResult(success=False, error=f"Failed to archive dialogue: {err}")


# =============================================================================
# Internal Helpers
# =============================================================================


def _parse_time(timestamp: str) -> int | None:
    """Parse HH:MM timestamp to minutes since midnight."""
    parts = timestamp.split(":")
    if len(parts) != 2:
        return None
    try:
        hours = int(parts[0])
        minutes = int(parts[1])
    except ValueError:
        return None
    return hours * 60 + minutes
